src/types.rs

```diff
@@ -370,9 +403,333 @@ impl From<ReturnCode> for u8 {
     }
 }
 
+/// RPC payload
+pub type RpcPayload<'a> = &'a [u8];
+
+#[derive(Debug, PartialEq)]
+/// SD payload
+pub struct SdPayload {
+    /// Flags
+    pub flags: SdFlags,
+    /// Entries
+    pub entries: Vec<SdEntry>,
+    /// Options
+    pub options: Vec<SdOption>,
+}
+
+impl SdPayload {
+    /// Length of the payload in bytes
+    pub(crate) fn len(&self) -> usize {
+        let mut len: usize = SD_HEADER_LENGTH;
+
+        len += 4;
+        len += self.entries_len();
+
+        len += 4;
+        len += self.options_len();
+
+        len
+    }
+
+    /// Length of the payload's entries in bytes
+    pub(crate) fn entries_len(&self) -> usize {
+        self.entries.len() * SD_ENTRY_LENGTH
+    }
+
+    /// Length of the payload's options in bytes
+    pub(crate) fn options_len(&self) -> usize {
+        let mut len: usize = 0;
+
+        for option in &self.options {
+            len += option.len();
+        }
+
+        len
+    }
+
+    /// Reboot flag
+    pub fn reboot_flag(&self) -> bool {
+        self.flags & 0x80 != 0x00
+    }
+
+    /// Unicast flag
+    pub fn unicast_flag(&self) -> bool {
+        self.flags & 0x40 != 0x00
+    }
+
+    /// Returns the associated options of an entry
+    pub fn options(&self, entry_index: usize) -> Vec<&SdEndpointOption> {
+        use SdEntry::*;
+        match self.entries.get(entry_index) {
+            Some(FindService(entry)) => self.referenced_options(&entry.options),
+            Some(OfferService(entry)) => self.referenced_options(&entry.options),
+            Some(SubscribeEventgroup(entry)) => self.referenced_options(&entry.options),
+            Some(SubscribeEventgroupAck(entry)) => self.referenced_options(&entry.options),
+            None => vec![],
+        }
+    }
```

Comment:
If this is something that is called often a iterator variant additionally to the owned version here could improve performance:

```rust
pub fn options_iter(&'a self, entries_index: usize) -> impl Iterator<Item = &'a SdEndpointOption>;
```

Reply:
?

Reply:
Instead of collecting something in a vector (which results in a allocation) you can return an iteratior that would need in that case just a index the current option. The `Iterator::Item` is `&SdEndpointOption`.  This more powerful than a collection.

Reply:
todo for further pr